pkg/mutation/system_test.go

```diff
@@ -301,25 +301,179 @@ func TestMutation(t *testing.T) {
 				t.Fatal(tc.tname, "Mutate failed unexpectedly", err)
 			}
 
-			accessor, err := meta.Accessor(toMutate)
-			if err != nil {
-				t.Fatal("Failed to get unstruct accessor", err)
-			}
-
-			newLabels := accessor.GetLabels()
+			newLabels := toMutate.GetLabels()
 
 			if !mutated {
-				t.Error(tc.tname, "Mutation not as expected", cmp.Diff(newLabels, tc.expectedLabels))
+				t.Error(tc.tname, "Mutation not as expected", cmp.Diff(tc.expectedLabels, newLabels))
 			}
 
-			if !cmp.Equal(newLabels, tc.expectedLabels) {
-				t.Error(tc.tname, "Mutation not as expected", cmp.Diff(newLabels, tc.expectedLabels))
+			if diff := cmp.Diff(tc.expectedLabels, newLabels); diff != "" {
+				t.Error(tc.tname, "Mutation not as expected", diff)
 			}
 
-			probe := c.orderedMutators[0].(*MockMutator) // fetching a mock mutator to check the number of iterations
+			probe := c.orderedMutators[0].(*fakeMutator) // fetching a mock mutator to check the number of iterations
 			if probe.MutationCount != tc.expectedIterations {
 				t.Error(tc.tname, "Expected %d  mutation iterations, got", tc.expectedIterations, tc.mutations[0].MutationCount)
 			}
 		})
 	}
 }
+
+func mustParse(s string) parser.Path {
+	p, err := parser.Parse(s)
+	if err != nil {
+		panic(err)
+	}
+	return p
+}
+
+func TestSystem_DontApplyConflictingMutations(t *testing.T) {
+	// Two conflicting mutators.
+	foo := &fakeMutator{
+		MID:    types.ID{Name: "foo"},
+		MPath:  mustParse("spec.containers[name: foo].image"),
+		GVKs:   []schema.GroupVersionKind{{Version: "v1", Kind: "Pod"}},
+		Labels: map[string]string{"active": "true"},
+	}
+	fooConflict := &fakeMutator{
+		MID:    types.ID{Name: "foo-conflict"},
+		MPath:  mustParse("spec.containers.image"),
+		GVKs:   []schema.GroupVersionKind{{Version: "v1", Kind: "Pod"}},
+		Labels: map[string]string{"active": "true"},
+	}
+
+	s := NewSystem()
+	err := s.Upsert(foo)
+	if err != nil {
+		t.Fatalf("got Upsert() error = %v, want <nil>", err)
+	}
+
+	// We can mutate objects before System is put in an inconsistent state.
+	u := &unstructured.Unstructured{}
+	gotMutated, gotErr := s.Mutate(u, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "billing"}})
+	if !gotMutated {
+		t.Errorf("got Mutate() = %t, want true", gotMutated)
```

Comment:
We should wrap these assertions in individual test names to disambiguate which line is failing.

See:

https://github.com/open-policy-agent/gatekeeper/blob/46e026a5aee204775a99294ce0f3e6b37effc39a/pkg/controller/constrainttemplate/constrainttemplate_controller_test.go#L217-L240

for an example

Reply:
Done.